File: python/d07_smart.py
# this approach works backwards from the goal, and continues only with the options that are possible

from typing import Literal

with open("../data/d07.txt") as f:
    lines = f.readlines()

# example:
# valid(v=123456, terms=[2, 617, 56])
# |-ADD-> valid(123400, [2, 617])
# |       |-ADD-> valid(122783, [2]) => ❌ 122783 != 2
# |       |-MUL-> valid(200, [2]) => ❌ 200 != 2
# |       |-CAT-> 🚫 impossible, since 123400 does not end in 56
# |-MUL-> 🚫 impossible, since 123400 % 617 != 0
# |-CAT-> valid(1234, [2, 617])
# |       |-ADD-> valid(1222, [2]) => ❌ 1222 != 2
# |       |-MUL-> valid(2, [2]) => ✅
# |       |-CAT-> valid(1234, [2]) => ❌ 1234 != 2


type Operator = Literal["add", "mul", "concat"]


def valid(v: int, terms: list[int], operators: list[Operator]) -> bool:
    #  print(f"valid({v}, {terms}, {operators})")
    if len(terms) == 1:
        return v == terms[0]

    next_terms, t = terms[:-1], terms[-1]

    def handle_op(op: Operator) -> bool:
        if op == "add":
            if v < t:
                return False
            return valid(v - t, next_terms, operators)
        elif op == "mul":
            if v % t != 0:
                return False
            return valid(v // t, next_terms, operators)
        elif op == "concat":
            q, r = divmod(v, 10 ** len(str(t)))
            if r != t:
                return False
            return valid(q, next_terms, operators)
        else:
            raise ValueError(f"invalid operator: {op}")

    return any(handle_op(op) for op in operators)


def is_line_valid(line: str, operators: list[Operator]) -> int:
    goal, rest = line.strip().split(": ")
    goal = int(goal)
    terms = [int(t) for t in rest.split()]

    if valid(goal, terms, operators):
        return goal
    return 0


print("p1", sum(is_line_valid(line, ["add", "mul"]) for line in lines))
print("p2", sum(is_line_valid(line, ["add", "mul", "concat"]) for line in lines))

File: rust/src/bin/d07_smart.rs
//! Smart solution that works backwards from the goal, and continues only with the options that are possible

fn main() {
    let input = include_str!("../../../data/d07.txt");
    let parsed_input = parse_input(input);
    println!("part one: {}", part_one(&parsed_input));
    println!("part two: {}", part_two(&parsed_input));
}

type ParsedInput = Vec<(usize, Vec<usize>)>;
fn parse_input(input: &str) -> ParsedInput {
    input
        .lines()
        .map(|line| {
            let (goal, rest) = line.split_once(": ").expect("invalid line");
            let goal = goal.parse::<usize>().expect("goal is not a number");
            let terms = rest
                .split_whitespace()
                .map(|t| t.parse::<usize>().expect("term is not a number"))
                .collect::<Vec<_>>();
            (goal, terms)
        })
        .collect::<Vec<_>>()
}

#[derive(Debug, Clone, Copy)]
enum Operator {
    Add,
    Mul,
    Concat,
}

fn valid(v: usize, terms: &[usize], term_idx: usize, operators: &[Operator]) -> bool {
    if term_idx == 0 {
        return terms[term_idx] == v;
    }

    let t = terms[term_idx];
    let next_idx = term_idx - 1;

    operators.iter().any(|op| match op {
        Operator::Add => {
            if v < t {
                return false;
            }
            valid(v - t, terms, next_idx, operators)
        }
        Operator::Mul => {
            if v % t != 0 {
                return false;
            }
            valid(v / t, terms, next_idx, operators)
        }
        Operator::Concat => {
            let next_power = get_next_power(t);
            let q = v / next_power;
            let r = v % next_power;
            if r != t {
                return false;
            }
            valid(q, terms, next_idx, operators)
        }
    })
}

const fn get_next_power(x: usize) -> usize {
    if x < 10 {
        10
    } else if x < 100 {
        100
    } else if x < 1000 {
        1000
    } else if x < 10000 {
        10000
    } else {
        panic!("x is too large");
    }
}

fn part_one(parsed_input: &ParsedInput) -> usize {
    let mut result = 0;
    let operators = [Operator::Add, Operator::Mul];
    for (goal, terms) in parsed_input {
        if valid(*goal, terms, terms.len() - 1, &operators) {
            result += goal;
        }
    }

    result
}

fn part_two(parsed_input: &ParsedInput) -> usize {
    let mut result = 0;
    let operators = [Operator::Add, Operator::Mul, Operator::Concat];
    for (goal, terms) in parsed_input {
        if valid(*goal, terms, terms.len() - 1, &operators) {
            result += goal;
        }
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    const TEST_INPUT: &str = include_str!("../../../test-data/d07.txt");

    #[test]
    fn test_part_one() {
        let parsed_input = parse_input(TEST_INPUT);
        assert_eq!(part_one(&parsed_input), 3749);
    }

    #[test]
    fn test_part_two() {
        let parsed_input = parse_input(TEST_INPUT);
        assert_eq!(part_two(&parsed_input), 11387);
    }
}
